### PR TITLE
Add BFloat16 dtype support for oneDNN Graph JIT fuser

### PR DESCRIPTION
**Pull request by sanchitintel**
## BFloat16 dtype support for faster inference with TorchScript using oneDNN Graph

Intel Xeon Cooper Lake platform & beyond support the `AVX512_BF16` ISA, which is essentially native BFloat16 support.
oneDNN Graph delivers high inference performance with BFloat16 on such machines.

While oneDNN Graph can still be used with BFloat16 on older machines that lack `avx512_bf16` ISA but support `avx512bw`, `avx512vl` & `avx512dq` ISAs, the BF16 performance on these older machines will be significantly poorer (probably even poorer than Float32), as they lack native BF16 support.

Currently, [AMP support for eager mode & JIT mode is divergent in PyTorch](https://github.com/pytorch/pytorch/issues/75956).
So, for using oneDNN Graph with BFloat16, eager-mode AMP should be leveraged by turning off AMP for JIT mode, using `torch._C._jit_set_autocast_mode(False)` in python code, so as to avoid conflicts.

Please use the following environment variable to view JIT logs -
`PYTORCH_JIT_LOG_LEVEL=">>graph_helper:>>graph_fuser:>>kernel:>>interface"`

## Changes being made in this PR
1. This PR does NOT change the `oneDNN` commit or the `ideep` files. While the `ideep` commit is being updated, only files pertaining to oneDNN Graph are being updated. oneDNN Graph is being upgraded to version 0.5.2 (alpha patch release 2).
To put things into perspective, `ideep` is a git submodule of PyTorch. `oneDNN Graph` is a git submodule of `ideep` (`ideep/mkl-dnn`), and oneDNN is a git submodule of oneDNN Graph (`ideep/mkl-dnn/third_party/oneDNN`).
2. Unit-tests are being updated. We now use the [existing dtypes decorator](https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_device_type.py#L123-L131).
3. Suggestions made by @eellison in the [FP32 PR](https://github.com/pytorch/pytorch/pull/68111#pullrequestreview-896719477) are being incorporated/addressed - 

| Action-item | Status |
| :---                                             |          ---: | 
|checkInputCompatibility follow up | Fixed |
|the mayConvertScalarInputToTensor logic we can consider | Added type promotion code |
|fix up fixConvOptionalBias| The current approach seems correct |
|Use opinfo tests| using dtypes decorator. Will use `OpInfo` in a subsequent PR, if that'd be possible. Should we create a list of ops from opDB that are supported by oneDNN Graph, and add it to `common_methods_invocations.py`? |
|inferDevice torch_check call | not necessary now, perhaps, as only CPU is supported, for now? We'd add it by the beta release of oneDNN Graph, though, so that by then, users might be able to use other fusers with oneDNN Graph (NNC/TensorExpr are already compatible with the oneDNN Graph fuser). We can still add it, if you'd insist. |
|not checking shapes of input mkldnn tensor to llga guard | Those checks should not be present because oneDNN Graph may use blocked or channels-last layout, so those strides would be different. They're only skipped if an LLGA subgraph's output is input to another LLGA subgraph, which enables LLGA to choose an optimal layout between them. |
|fix test failures with respect to unsupported inputs | We'll address them with the upcoming release of oneDNN Graph beta version|

4. More PyTorch ops are being been mapped to oneDNN Graph

## Example of using oneDNN Graph with BFloat16

```python
# Assuming we have a model of the name 'model'

example_input = torch.rand(1, 3, 224, 224)

# enable oneDNN Graph
torch.jit.enable_onednn_fusion(True)
# Disable AMP for JIT
torch._C._jit_set_autocast_mode(False)
with torch.no_grad(), torch.cpu.amp.autocast():
    model = torch.jit.trace(model, (example_input))
    model = torch.jit.freeze(model)
     # 2 warm-ups (2 for tracing/scripting with an example, 3 without an example)
    model(example_input)
    model(example_input)
    
    # speedup would be observed in subsequent runs.
    model(example_input)
``` 

## TorchBench based Benchmarks
**URL:** https://github.com/sanchitintel/benchmark/tree/onednn_graph_benchmark (instructions present at URL).
**Batch-size(s):** TorchBench-default for each model
**Baseline :** PyTorch JIT OFI FP32
**Machine:** Intel(R) Xeon(R) Platinum 8371HC (Cooper Lake)
**Sockets used**: 1
**Number of cores on one socket**: 26
Intel OpenMP & tcmalloc were preloaded

#### Benchmark results with single thread
| name                                             | latency of PyTorch JIT OFI FP32 (s) |   Latency of oneDNN Graph BF16 (s) |   % change |
| :---                                             |          ---: |            ---: |       ---: |
| test_eval[alexnet-cpu-jit]                       |      1.063851 |        0.509820 |     -52.1% |
| test_eval[mnasnet1_0-cpu-jit]                    |      0.218435 |        0.107100 |     -51.0% |
| test_eval[mobilenet_v2-cpu-jit]                  |      0.114467 |        0.058359 |     -49.0% |
| test_eval[mobilenet_v3_large-cpu-jit]            |      0.233873 |        0.117614 |     -49.7% |
| test_eval[resnet18-cpu-jit]                      |      0.160584 |        0.075854 |     -52.8% |
| test_eval[resnet50-cpu-jit]                      |      1.652846 |        0.713373 |     -56.8% |
| test_eval[resnext50_32x4d-cpu-jit]               |      0.471174 |        0.209431 |     -55.6% |
|test_eval[shufflenet_v2_x1_0-cpu-jit] | 0.310306 | 0.167090 | -46.2% |
| test_eval[squeezenet1_1-cpu-jit]                 |      0.161247 |        0.045684 |     -71.7% |
| test_eval[timm_efficientnet-cpu-jit]             |      1.643772 |        0.800099 |     -51.3% |
| test_eval[timm_regnet-cpu-jit]                   |      5.732272 |        2.333417 |     -59.3% |
| test_eval[timm_resnest-cpu-jit]                  |      1.366464 |        0.715252 |     -47.7% |
| test_eval[timm_vision_transformer-cpu-jit]       |      0.508521 |        0.271598 |     -46.6% |
| test_eval[timm_vovnet-cpu-jit]                   |      2.756692 |        1.125033 |     -59.2% |
| test_eval[vgg16-cpu-jit]                         |      0.711533 |        0.312344 |     -56.1% |

#### Benchmark results with 26 threads:
| name                                             | latency of PyTorch JIT OFI FP32 (s) |   Latency of oneDNN Graph BF16 (s) |   % change |
| :---                                             |          ---: |            ---: |       ---: |
| test_eval[alexnet-cpu-jit]                       |      0.062871 |        0.034198 |     -45.6% |
| test_eval[mnasnet1_0-cpu-jit]                    |      0.022490 |        0.008172 |     -63.7% |
| test_eval[mobilenet_v2-cpu-jit]                  |      0.012730 |        0.005866 |     -53.9% |
| test_eval[mobilenet_v3_large-cpu-jit]            |      0.025948 |        0.010346 |     -60.1% |
| test_eval[resnet18-cpu-jit]                      |      0.011194 |        0.005726 |     -48.9% |
| test_eval[resnet50-cpu-jit]                      |      0.124662 |        0.045599 |     -63.4% |
| test_eval[resnext50_32x4d-cpu-jit]               |      0.034737 |        0.015214 |     -56.2% |
|test_eval[shufflenet_v2_x1_0-cpu-jit] | 0.028820 | 0.012517 | -56.6% |
| test_eval[squeezenet1_1-cpu-jit]                 |      0.012557 |        0.003876 |     -69.1% |
| test_eval[timm_efficientnet-cpu-jit]             |      0.203177 |        0.051879 |     -74.5% |
| test_eval[timm_regnet-cpu-jit]                   |      0.452050 |        0.151113 |     -66.6% |
| test_eval[timm_resnest-cpu-jit]                  |      0.117072 |        0.052848 |     -54.9% |
| test_eval[timm_vision_transformer-cpu-jit]       |      0.046048 |        0.023275 |     -49.5% |
| test_eval[timm_vovnet-cpu-jit]                   |      0.213187 |        0.077482 |     -63.7% |
| test_eval[vgg16-cpu-jit]                         |      0.044726 |        0.021998 |     -50.8% |

### CODE REVIEW

File: test/test_jit_llga_fuser.py
Comment (sanchitintel) [inline test/test_jit_llga_fuser.py:25]: @chunyuan-w, reusing `has_bf16_support` from `test_mkldnn.py` resulted in 2 CI checks failing (1 rocm & 1 CUDA),
so I readded this one instead, as all CI checks had passed with this approach yesterday. Thanks!

**Review by sanchitintel — COMMENTED** [review]
@XiaobingSuper, added you as a reviewer, because this PR uses a distinct ideep commit (same oneDNN commit in `ideep/mkl-dnn/third_party/oneDNN` & doesn't change any `ideep` file, but only changes oneDNN Graph files (`ideep/mkl-dnn`). Thanks!

**Comment by sanchitintel** [comment]
Hi @eellison, thanks a lot for your help earlier with integrating oneDNN Graph with PyTorch!
Just FYI, here's the status of addressing/incorporating [your suggestions](https://github.com/pytorch/pytorch/pull/68111#pullrequestreview-896719477)  -

| Concern                                             | Status |
| :---                                             |          ---: | 
|checkInputCompatibility follow up | Fixed |
|the mayConvertScalarInputToTensor logic we can consider | Added type promotion code |
|fix up fixConvOptionalBias| The current approach seems correct |
|Use opinfo tests| using dtypes decorator. Will use `OpInfo` in a subsequent PR, if that'd be possible. Should we create a list of ops from opDB that are supported by oneDNN Graph, and add it to `common_methods_invocations.py`? |
|inferDevice torch_check call | not necessary now, perhaps, as only CPU is supported, for now? We'd add it by the beta release of oneDNN Graph, though, so that by then, users might be able to use other fusers with oneDNN Graph (NNC/TensorExpr are already compatible with the oneDNN Graph fuser). We can still add it, if you'd insist. |
|not checking shapes of input mkldnn tensor to llga guard | Those checks should not be present because oneDNN Graph may use blocked or channels-last layout, so those strides would be different. They're only skipped if an LLGA subgraph's output is input to another LLGA subgraph, which enables LLGA to choose an optimal layout between them. |
|fix test failures with respect to unsupported inputs | We'll address them with the upcoming release of oneDNN Graph beta version|

Please go through them, whenever possible. Thanks!

**Comment by frank-wei** [comment]
cc @malfet , do we remove the choice of importing the diff into internal?

**Comment by malfet** [comment]
> cc @malfet , do we remove the choice of importing the diff into internal?

No, but the plugin has been finicky throughout the day

File: test/test_jit_llga_fuser.py
Comment (sanchitintel) [reply test/test_jit_llga_fuser.py]: Uh, oh! Not sure why I did so at the time. Thanks for catching this! :)

File: torch/csrc/jit/codegen/onednn/operator.h
Comment (sanchitintel) [reply torch/csrc/jit/codegen/onednn/operator.h]: Thank you! That's a good idea!

File: torch/csrc/jit/codegen/onednn/LlgaTensorImpl.cpp
Comment (sanchitintel) [reply torch/csrc/jit/codegen/onednn/LlgaTensorImpl.cpp]: Thanks for pointing this out!
`shufflenet_v2_x1_0` has a `prim::Constant` of value 2 with dtype Long. oneDNN Graph doesn't support ScalarType::Long (`int64_t`), so we should not support such models.

File: test/test_jit_llga_fuser.py
Comment (chunyuan-w) [inline test/test_jit_llga_fuser.py:37]: Is this line still needed since you've defined `self.original_autocast_mode` later on?

Comment (sanchitintel) [reply test/test_jit_llga_fuser.py:37]: Thanks for catching this!

File: torch/csrc/jit/codegen/onednn/graph_helper.cpp
Comment (sanchitintel) [reply torch/csrc/jit/codegen/onednn/graph_helper.cpp:378]: Uh, oh!

**Comment by sanchitintel** [comment]
CI system is down for some CI checks, and the failing issues are unrelated. Thanks! 

image

**Comment by sanchitintel** [comment]
The CI failure is not an actual failure but is an indicator that the commit used by Meta-internal CI check is divergent from the HEAD commit of this PR's branch. Thanks!

File: torch/csrc/jit/codegen/onednn/LlgaTensorImpl.cpp
Comment (sanchitintel) [reply torch/csrc/jit/codegen/onednn/LlgaTensorImpl.cpp:97]: Added a [UT to demonstrate.](https://github.com/pytorch/pytorch/pull/85591/commits/64da70c11a76cd59a6a415dfb240d8f50d10b1b8) Thanks!

**Comment by sanchitintel** [comment]
CI failures are unrealted. Thanks!

eg. 
image

**Review by jgong5 — APPROVED** [review]

**Comment by sanchitintel** [comment]
Hi @frank-wei, this PR wouldn't have any further revisions.
Whenever you can, please re-import it for Meta-internal CI checks. Thanks!

**Comment by sanchitintel** [comment]
Thank you for re-importing the PR, @frank-wei! 
Whenever possible, can you please check if all Meta-internal CI checks had passed for this PR?
If they did not, we'd request your help with resolving any potential issues next month.
But if they passed, can you please help land this PR today? Thanks!

**Comment by frank-wei** [comment]
Internal tests are passed. Merge this PR now.

**Comment by frank-wei** [comment]
Need @malfet approval. Mine does not work on this PR.

**Comment by malfet** [comment]
@frank-wei not sure what do you mean by that? Why yours would not work?

**Review by frank-wei — APPROVED** [review]

**Comment by frank-wei** [comment]
> @frank-wei not sure what do you mean by that? Why yours would not work?

sorry, let me try again to merge this PR to see if it is due to lack of my approval in previous try.

**Comment by frank-wei** [comment]
@sanchitintel  could you please rebase this PR? thanks.

**Comment by sanchitintel** [comment]
Hi @frank-wei, thanks again for your help!
We've modified the UTs in this PR to work around the existing memory-leak issue of JIT traced models' objects' memory not being freed.
It's a small change that runs some UTs in separate processes, so that the memory used by the main UT process does not cumulatively become large enough for the Linux kernel's out-of-memory killer to be invoked on `linux.2xlarge` CI runners, which have 16 GB RAM.

As such, this change may not necessitate reimporting the PR for internal CI checks.
Whenever possible, please re-import or merge this PR. Thanks!

**Review by jgong5 — APPROVED** [review]

**Review by chunyuan-w — APPROVED** [review]

**Review by frank-wei — APPROVED** [review]
LGTM!

**Comment by sanchitintel** [comment]
Thanks again, @frank-wei! It looks like it'd have to be imported again before merging.

**Comment by sanchitintel** [comment]
Hi @malfet, please help me with a query about the `pytorchmergebot` rule for not merging a PR if the last push to it was more than 3 days ago - my last push to this PR was on Oct 10. Does this rule mean that the PR must be merged on or before Oct 13, or the PR would've to be rebased on Oct 14 (& subsequently reimported as well - since `pytorchmergebot` didn't merge this PR last time because of the `Meta Internal-Only Changes Check` due to rebased commit being different than the commit for which Meta-internal CI checks were run, so I'm assuming it'd do so again, if it comes to that)?

Thanks!

**Comment by frank-wei** [comment]
I feel like once we imported this PR, it always has to sync internally whenever we want to merge it from outside. This could get improved since once the PR is on diff train, the diff is supposed to re-created internally.
For stale state check, I do not have clear answer. Better check with @malfet

**Comment by sanchitintel** [comment]
Thanks for your inputs, @frank-wei! :)

If the answer to my query about the pytorchmergebot stale check is yes, then IMHO, the 3 day period seems a bit small, as it might require Meta engineers to frequently re-import PRs. But on the other hand, I guess it does seem to help keep the trunk CI greener, so maybe it was determined as being worth the trade-off. :)

**Comment by sanchitintel** [comment]
Thanks again, @frank-wei! :)

**Comment by malfet** [comment]
Note to internal oncall: this PR updates ideep, it needs to be kept in sync internally.
cc: @weiwangmeta

**Comment by sanchitintel** [comment]
> this PR updates ideep, it needs to be kept in sync internally.

Hi @weiwangmeta, we'd be submitting a PR to update ideep again tonight with a new oneDNN version. I'll add you as a reviewer in that PR. Thanks!

**Pull request closed without merging**
